packages/contractkit/src/identity/offchain/schemas/utils.ts

```diff
@@ -0,0 +1,348 @@
+
+  await Promise.all(
+    // here we encrypt the key to ourselves so we can retrieve it later
+    [wrapper.self, ...toAddresses].map(async (toAddress) =>
+      distributeSymmetricKey(wrapper, dataPath, fetchKey.result, toAddress)
+    )
+  )
+}
+
+/**
+ * Reads and decrypts a symmetric key that has been encrypted to your
+ * data encryption key.
+ *
+ * @param wrapper the offchain data wrapper
+ * @param dataPath path to where the encrypted data is stored. Used to derive the key location
+ * @param senderAddress the address that encrypted this key to you
+ */
+const readSymmetricKey = async (
+  wrapper: OffchainDataWrapper,
+  dataPath: string,
+  senderAddress: string
+): Promise<Result<Buffer, SchemaErrors>> => {
+  const accounts = await wrapper.kit.contracts.getAccounts()
+  const wallet = wrapper.kit.getWallet()
+  const [readerPubKey, senderPubKey] = await Promise.all([
+    accounts.getDataEncryptionKey(wrapper.self),
+    accounts.getDataEncryptionKey(senderAddress),
+  ])
+
+  if (readerPubKey === null) {
+    return Err(new UnavailableKey(senderAddress))
+  }
+
+  const readerPublicKeyAddress = publicKeyToAddress(readerPubKey)
+  if (!wallet.hasAccount(readerPublicKeyAddress)) {
+    return Err(new UnavailableKey(readerPublicKeyAddress))
+  }
+
+  const sharedSecret = await wallet.computeSharedSecret(readerPublicKeyAddress, senderPubKey)
+  const computedDataPath = getCiphertextLabel(
+    `${dataPath}.key`,
+    sharedSecret,
+    senderPubKey,
+    readerPubKey
+  )
+  const encryptedPayload = await wrapper.readDataFromAsResult(
+    senderAddress,
+    (buf) => buildEIP712TypedData(wrapper, computedDataPath, buf),
+    computedDataPath
+  )
+
+  if (!encryptedPayload.ok) {
+    return Err(new OffchainError(encryptedPayload.error))
+  }
+
+  const payload = await wallet.decrypt(readerPublicKeyAddress, encryptedPayload.result)
+  return Ok(payload)
+}
+
+/**
+ * Reads and decrypts a payload that has been encrypted to your data encryption key. Will
+ * resolving the symmetric key used to encrypt the payload.
+ *
+ * @param wrapper the offchain data wrapper
+ * @param dataPath path to where the encrypted data is stored. Used to derive the key location
+ * @param senderAddress the address that encrypted this key to you
+ */
+export const readEncrypted = async (
+  wrapper: OffchainDataWrapper,
+  dataPath: string,
+  senderAddress: string
+): Promise<Result<Buffer, SchemaErrors>> => {
+  const encryptedPayloadPath = `${dataPath}.enc`
+  const [payload, key] = await Promise.all([
+    wrapper.readDataFromAsResult(
+      senderAddress,
+      (buf) => buildEIP712TypedData(wrapper, encryptedPayloadPath, buf),
+      encryptedPayloadPath
+    ),
+    readSymmetricKey(wrapper, dataPath, senderAddress),
+  ])
+
+  if (!payload.ok) {
+    return Err(new OffchainError(payload.error))
+  }
+  if (!key.ok) {
+    return Err(key.error)
+  }
+
+  if (key.result.length !== KEY_LENGTH) {
+    return Err(new InvalidKey())
+  }
+
+  return Ok(
```

Comment:
We should add a minimum length check before this point to make sure that the payload is at least `IV_LENGTH` bytes.